### PR TITLE
Pass Variable into Caffe2 ops, by requiring that the Variable doesn't require grad

### PR DESCRIPTION
**Pull request by yf225**
As part of the Variable/Tensor merge, we want to be able to pass Variables into Caffe2 without doing extra shallow copy, to improve performance and also allow for in-place mutations in Caffe2 ops. There are a few approaches outlined in https://github.com/pytorch/pytorch/pull/22418, and this PR is the chosen approach.

Specifically, we can have the assumption that we won't be connecting autograd to C2 gradients at any point (as it's too tricky and not that useful). Therefore, we can pass Variable into Caffe2 ops by requiring that all Variables in Caffe2 don't require grad. For code paths in Caffe2 that might potentially track gradients (e.g. `ScriptModuleOp` and `call_caffe2_op_from_c10`), we use the `torch::NoGradGuard` to make sure gradients are not tracked.

This supersedes https://github.com/pytorch/pytorch/pull/22418.

### CODE REVIEW

File: caffe2/core/tensor.h
Comment (smessmer) [inline caffe2/core/tensor.h]: like above, this should probably be replaced with a check that `requires_grad=False` instead of just deleted.

Comment (yf225) [reply caffe2/core/tensor.h]: fixed.

File: torch/csrc/api/include/torch/grad_mode.h
Comment (smessmer) [inline torch/csrc/api/include/torch/grad_mode.h]: nit: can be final

File: torch/csrc/autograd/python_variable.cpp
Comment (smessmer) [inline torch/csrc/autograd/python_variable.cpp]: You might be able to avoid this check if you instantiate the outputs in the wrapper correctly.
Currently, the `outputs` variable there is just instantiated to be uninitialized `at::Tensor`s which means they're not variables.

**Review by smessmer — APPROVED** [review]
Please fix the output situation. Otherwise looks good.

File: caffe2/core/export_caffe2_op_to_c10.h
Comment (gchanan) [inline caffe2/core/export_caffe2_op_to_c10.h]: a comment would be great.

Comment (yf225) [reply caffe2/core/export_caffe2_op_to_c10.h]: Fixed.

File: torch/csrc/autograd/python_variable.cpp
Comment (yf225) [reply torch/csrc/autograd/python_variable.cpp]: @smessmer I think we might still need this check because of the following use case: https://github.com/pytorch/pytorch/blob/10c4b98ade8349d841518d22f19a653a939e260c/caffe2/python/workspace_test.py#L296-L298
Here, the tensor `foo` is created from Caffe2, and it's not a Variable. When `FetchTorch` is run which goes through the `_wrap_tensor_impl` code path, it will find that `foo` is not a Variable, and it will throw error if we directly run `torch::autograd::Variable(tensor)`. Hence I think we still need the check here and do the wrapping when necessary.

**Comment by yf225** [comment]
Since Caffe2 tensor's `enforce_invariants()` method needs to check `torch::autograd::GradMode::is_enabled()` (lives in `torch/csrc/autograd`), and Caffe2 tensor lives in Caffe2 core, @ljk53 suggested that we should move `torch::autograd::GradMode` from `torch/csrc/autograd` to ATen core, so that core code doesn't need to depend on non-core code, which can prevent build structure headache. PR for moving `GradMode` is at https://github.com/pytorch/pytorch/pull/18573.

File: caffe2/core/tensor.cc
Comment (gchanan) [inline caffe2/core/tensor.cc]: you don't need to check is_variable again -- if it got to the or, you know it's a variable.

File: caffe2/core/tensor.cc
Comment (gchanan) [inline caffe2/core/tensor.cc]: we don't have GradMode on mobile or something?  That seems...annoying to keep trac kof.

Comment (yf225) [reply caffe2/core/tensor.cc]: @ljk53 told me that `thread_local` should work on mobile. After https://github.com/pytorch/pytorch/pull/18573 I will try to remove the spacial casing and see if we run into any build error.

File: torch/csrc/autograd/python_variable.cpp
Comment (smessmer) [reply torch/csrc/autograd/python_variable.cpp]: makes sense, we still need it. It still makes sense to create the preallocated output tensors as variables though, because that's a potential behavior change we want to be sure caffe2 ops can handle.

Comment (yf225) [reply torch/csrc/autograd/python_variable.cpp]: I think it might be a bit tricky to change the `outputs` from `at::Tensor` to `torch::autograd::Variable`, because that would mean that `caffe2/core/export_caffe2_op_to_c10.h` needs to depend on `torch/csrc/autograd/variable.h`, which is not part of the core code.

I think an easier approach might be to wait until Variable/Tensor merge is complete, at which point the `torch::autograd::Variable` class is removed and `at::Tensor` is a Variable automatically (with no autograd metadata), and therefore no change is needed for the `outputs` construction code in `caffe2/core/export_caffe2_op_to_c10.h`.

**Pull request closed without merging**
